### PR TITLE
refactor: improve performance of wrap_generic_callable

### PR DESCRIPTION
**Pull request by lavarou**
`nr_php_wrap_generic_callable` is used to create transient wraprecs for callables which **_don't require to be named_**. This assumption makes it possible to stay DRY and reuse `nr_php_zval_to_function` in lieu of complex pattern matching for the type of callable. This reduces agent's performance overhead as it now uses `nr_php_wrap_callable` always instead of sometimes using `nr_php_wrap_user_function_with_transience(IS_TRANSIENT)`. This is an alternative to #665 which obsoletes and removes #679 and makes #770 no longer needed.

TL:DR; The agent's performance overhead when `nr_php_wrap_user_function_with_transience(IS_TRANSIENT)` is used comes from the need to walk the linked list of named wraprecs in `nr_php_add_custom_tracer_named` [here](https://github.com/newrelic/newrelic-php-agent/blob/5a1ef2bcd7ed1e5ffde87c9175cce8e41abb8268/agent/php_user_instrument.c#L413-L426).

### CODE REVIEW

**Review by ZNeumann — APPROVED** [review]

File: agent/php_wrapper.c [hunk 1]
```diff
@@ -7,20 +7,47 @@
...
+    case IS_OBJECT:
+      return "IS_OBJECT";
+
+    case IS_ARRAY:
+      return "IS_ARRAY";
+
+    default:
+      return "UNKNOWN";
+  }
+}
```
Comment (lavarou) [inline agent/php_wrapper.c:45]: I'm not sure about this...

Comment (ZNeumann) [reply agent/php_wrapper.c:45]: I like it alright, though I think it would be beneficial to rename the function. With its current name, I would assume it is provided by the PHP internals API. It might also be worth mentioning that this is (I think) currently only used for logging. Maybe put "log" somewhere in the name so that somebody has to justify a rename for a more important use-case.

**Review by ZNeumann — APPROVED** [review]

File: agent/php_wrapper.c (hunk 1 above)
Comment (lavarou) [reply agent/php_wrapper.c:45]: I wonder if it might be better to make `nr_format_zval_for_debug` function external to `agent/php_execute.c`. I sometimes wish it was as it makes debugging so much easier...

Comment (bduranleau-nr) [reply agent/php_wrapper.c:45]: I agree with @ZNeumann's comment regarding the naming convention of this function.

Comment (lavarou) [reply agent/php_wrapper.c:45]: @ZNeumann @bduranleau-nr What do you think about the idea of making `nr_format_zval_for_debug` function external to `agent/php_execute.c` and using it instead of adding `Z_TYPE_STR`?

Comment (ZNeumann) [reply agent/php_wrapper.c:45]: works for me

Comment (bduranleau-nr) [reply agent/php_wrapper.c:45]: agree 👍

**Review by bduranleau-nr — APPROVED** [review]

File: agent/php_wrapper.c (hunk 1 above)
Comment (lavarou) [reply agent/php_wrapper.c]: @bduranleau-nr, @ZNeumann: See e3938763a55965361a99778e52e94578dc382b8f.

File: agent/php_wrapper.c [hunk 2]
```diff
@@ -68,79 +59,22 @@ nruserfn_t* nr_php_wrap_callable(zend_function* callable,
...
-  if (NULL != name) {
-    nrl_verbosedebug(NRL_INSTRUMENT,
-                   "Failed to wrap callable: %s", ZEND_STRING_VALUE(name));
-  } else {
-    nrl_verbosedebug(NRL_INSTRUMENT, "Failed to wrap callable with unknown name");
+  if (nrl_should_print(NRL_VERBOSEDEBUG, NRL_INSTRUMENT)) {
+    char strbuf[NR_WRAPPER_DEBUG_STRBUFSZ];
+    nr_format_zval_for_debug(callable, strbuf, 0, NR_WRAPPER_DEBUG_STRBUFSZ - 1,
+                             0);
+    nrl_verbosedebug(NRL_INSTRUMENT, "Failed to cast to callable zval=%s",
```
Comment (mfulb) [inline agent/php_wrapper.c:75]: Could this possibly output a string value which shouldn't appear in a log file?

File: tests/integration/frameworks/drupal/skipif.inc [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?php
+/*
+ * Copyright 2020 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+if (version_compare(PHP_VERSION, "8.2", ">=")) {
```
Comment (mfulb) [inline tests/integration/frameworks/drupal/skipif.inc:7]: Is this related to the changes in this PR?

Comment (lavarou) [reply tests/integration/frameworks/drupal/skipif.inc:7]: WordPress and Drupal instrumentation is using code paths modified by this PR. There already existed WordPress integration tests. However there were no Drupal integration tests. This PR adds Drupal integration tests to increase coverage of the Drupal instrumentation and code paths modified by this PR.

Comment (mfulb) [reply tests/integration/frameworks/drupal/skipif.inc:7]: Is there a similar Drupal test for PHP >= 8.2?

Comment (mfulb) [reply tests/integration/frameworks/drupal/skipif.inc:7]: After discussion it was determined the test for PHP >= 8.2 is because the `invoke_all` Drupal API call is not available in Drupal versions which run on PHP >= 8.2.  The test using `invoke_all_with` is run for those PHP versions.

**Review by mfulb — APPROVED** [review]

File: tests/integration/frameworks/drupal/skipif.inc (hunk 3 above)
Comment (lavarou) [reply tests/integration/frameworks/drupal/skipif.inc:7]: Yes there is. Older version of Drupal (the one that does not run on PHPs 8.2+) use one method of invoking custom plugins and this method is tested with `tests/integration/frameworks/drupal/test_module_invoke_all.php`. Newer version of Drupal (the one that runs on PHPs 7.4+) use another method of invoking custom plugins and this method is tested with `tests/integration/frameworks/drupal/test_invoke_all_with.php`.
